### PR TITLE
Additional searchers

### PR DESCRIPTION
**Pull request by samihamdan**
Implementing #98 
Adding: 
* registering as for transformers 
* model_selection as folder instead of one file (similar to scikit-learn)

### CODE REVIEW

File: docs/changes/latest.inc [hunk 1]
```diff
@@ -21,6 +21,8 @@ Enhancements
 
 - Update docs to make it more uniform (by `Kaustubh Patil`_). 
 
+- Update model_selection and add capabilities to register searchers (by `Sami Hamdan`_).
```
Comment (fraimondo) [inline docs/changes/latest.inc:24]: If with `model_selection` you refer to a module, make it a link.

File: julearn/model_selection/available_searchers.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+        * None : overwriting is possible, but warns the user
+        * True : overwriting is possible without any warning
+        * False : overwriting is not possible, error is raised instead
+
+    """
+    if name in list_searchers():
+        if overwriting:
+            _available_searchers[name] = searcher
+            logger.info(f'registering searcher names {name}')
```
Comment (fraimondo) [inline julearn/model_selection/available_searchers.py:70]: The message is not correct. Maybe `f'Registering new searcher: {name}'`

File: julearn/model_selection/available_searchers.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+            logger.info(f'registering searcher names {name}')
+
+        else:
+            raise_error(
+                f'searcher named {name} already exists and '
+                'overwriting is set to False, therefore you cannot overwrite '
+                'existing searchers. '
+                'Set `overwriting=True` in case you want to '
+                'overwrite existing scorers.'
```
Comment (fraimondo) [inline julearn/model_selection/available_searchers.py:87]: `"scorers"`

File: julearn/model_selection/available_searchers.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            )
+    else:
```
Comment (fraimondo) [inline julearn/model_selection/available_searchers.py:89]: There's no need for an else clause here.
```Python
if name in _available searchers:
    if overwrite is None:
        # Warn
    elif overwrite is False:
        # Error
logger.info(f'Registering new searcher: {name}')
_available_searchers[name] = searcher
```

File: julearn/model_selection/available_searchers.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    if name not in _available_searchers:
+        raise_error(
+            f'The specified searcher ({name}) is not available. '
+            f'Valid options are: {list(_available_searchers.keys())}'
+        )
+    out = _available_searchers[name]
+    return out
+
+
+def register_searcher(name, searcher, overwriting=None):
```
Comment (fraimondo) [inline julearn/model_selection/available_searchers.py:46]: Parameter should be called `overwrite`.  (maybe it's also like that in all the other `register`, should be changed).

File: julearn/model_selection/tests/test_available_searchers.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import pytest
+from sklearn.model_selection._search_successive_halving import (
```
Comment (fraimondo) [inline julearn/model_selection/tests/test_available_searchers.py:2]: Don't see this file. And why is this a private module (`_search_succesive_halving`)

File: julearn/model_selection/tests/test_available_searchers.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import pytest
```
Comment (fraimondo) [inline julearn/model_selection/tests/test_available_searchers.py:1]: Missing
```Python
# Authors: Sami Hamdan <s.hamdan@fz-juelich.de>
# License: AGPL
```

File: julearn/model_selection/available_searchers.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
+from copy import deepcopy
```
Comment (fraimondo) [inline julearn/model_selection/available_searchers.py:1]: Missing
```Python
# Authors: Sami Hamdan <s.hamdan@fz-juelich.de>
# License: AGPL
```

File: julearn/model_selection/__init__.py [hunk 9]
```diff
@@ -0,0 +1,3 @@
+from . cv import StratifiedBootstrap
```
Comment (fraimondo) [inline julearn/model_selection/__init__.py:1]: Missing
```Python
# Authors: Sami Hamdan <s.hamdan@fz-juelich.de>
# License: AGPL
```

**Review by fraimondo — CHANGES_REQUESTED** [review]

File: julearn/model_selection/tests/test_available_searchers.py (hunk 7 above)
Comment (samihamdan) [reply julearn/model_selection/tests/test_available_searchers.py:4]: Added both of us: as we did for all the other files in julearn (except for the examples)

File: julearn/model_selection/available_searchers.py (hunk 8 above)
Comment (samihamdan) [reply julearn/model_selection/available_searchers.py:4]: Added both of us: as we did for all the other files in julearn (except for the examples)

File: julearn/model_selection/__init__.py (hunk 9 above)
Comment (samihamdan) [reply julearn/model_selection/__init__.py:4]: Added both of us: as we did for all the other files in julearn (except for the examples)

File: docs/changes/latest.inc (hunk 1 above)
Comment (samihamdan) [reply docs/changes/latest.inc:24]: Is this correct: 
 :mod:\`.model_selection\`

File: julearn/model_selection/available_searchers.py (hunk 5 above)
Comment (samihamdan) [reply julearn/model_selection/available_searchers.py:49]: I changed for searchers in this PR. 
I will do the same for scorers in here #111 .
For transformers and estimators we will need another PR.

File: julearn/model_selection/tests/test_available_searchers.py (hunk 6 above)
Comment (samihamdan) [reply julearn/model_selection/tests/test_available_searchers.py]: I did not find a not used Searcher except for `HalvingGridSearchCV` and I was not able to import it directly from sklearn.model_selection` locally. This is why I did it this way. 
But I exchanged it not with normal `GridSearchCV` and the test work this way as well.
Seems like I over-complicated it before.

File: docs/changes/latest.inc (hunk 1 above)
Comment (fraimondo) [reply docs/changes/latest.inc]: Would be something like that. Does it work?

**Review by fraimondo — APPROVED** [review]

**Pull request merged**